src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,168 @@
+                unlinkItem(node);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<E> {
+        private E item;
+        private Node<E> next;
+        private Node<E> prev;
+
+        Node(Node<E> prev, E item, Node<E> next) {
+            this.prev = prev;
+            this.item = item;
+            this.next = next;
+        }
+    }
+
+    private T unlinkItem(Node<T> foundNode) {
+        final T removeElement = foundNode.item;
+        Node<T> next = foundNode.next;
+        Node<T> prev = foundNode.prev;
+
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            foundNode.prev = null;
+        }
+
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            foundNode.next = null;
+        }
+
+        foundNode.item = null;
+        size--;
+        return removeElement;
+    }
+
+    private void linkBeforeItem(T item, Node<T> foundNode) {
+        Node<T> prevNode = foundNode.prev;
+        Node<T> newNode = new Node<>(prevNode, item, foundNode);
+        foundNode.prev = newNode;
+        if (prevNode == null) {
+            first = newNode;
+        } else {
+            prevNode.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkedLastItem(T value) {
+        Node<T> tempLast = last;
+        Node<T> newNode = new Node<>(tempLast, value, null);
+        last = newNode;
+        if (tempLast == null) {
+            first = newNode;
+        } else {
+            tempLast.next = newNode;
+        }
+        size++;
+    }
+
+    Node<T> findNodeByIndex(int index) {
+        Node<T> fondNode;
+        if (index < (size - 1)) {
+            fondNode = first;
+            for (int i = 0; i < index; i++) {
+                fondNode = fondNode.next;
+            }
+            return fondNode;
+        } else {
+            fondNode = last;
+            for (int i = size - 1; i > index; i--) {
+                fondNode = fondNode.prev;
+            }
+            return fondNode;
+        }
+    }
+
+    private void checkElementIndex(int index) {
+        if (!isElementIndex(index)) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of range!");
+        }
+    }
+
+    private void checkPositionIndex(int index) {
+        if (!isPositionIndex(index)) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of range!");
+        }
+    }
+
+    private boolean isElementIndex(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private boolean isPositionIndex(int index) {
+        return index >= 0 && index <= size;
```

Comment:
We can also remove the boolean returning methods and inline them, since we are not using them anywhere else
and try not to negate the statements, and form them so that they would work without `!`